presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util.maps;
+
+import com.google.common.base.Equivalence;
+import org.intellij.lang.annotations.Flow;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * A Map decorator allowing for overriding the equivalence relation
+ * for the underlying Map's keys.
+ * <p>
+ * EquivalenceMap instances retain iteration order of the underlying Map.
+ * Collections returned from #entrySet() and #keySet() are unmodifiable
+ * views.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class EquilvalenceMap<K, V>
+        implements Map<K, V>
+{
+    private final Map<Equivalence.Wrapper<K>, V> delegate;
+    private final Equivalence<Object> equivalenceRelation;
+
+    public static <K, V> EquilvalenceMap<K, V> on(MapSupplier supplier, Equivalence<Object> equilvalenceRelation)
+    {
+        return new EquilvalenceMap<>(supplier, equilvalenceRelation);
+    }
+
+    protected EquilvalenceMap(MapSupplier supplier, Equivalence<Object> equivalenceRelation)
+    {
+        this.delegate = supplier.supply();
```

Comment:
So that the user does not have to know about the key type in the wrapped Map. Otherwise the constuctor would read:

`EquilvalenceMap(Map<Equivalence.Wrapper<K>, V> delegate, Equivalence<Object> equivalenceRelation)`.

The MapSupplier infers the type. And, thanks to the MapSupplier not knowing the key (and value) type in advance, there's less risk that users will pass a non-empty Map.